File: algorithms/2. Add Two Numbers/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        carry = 0
        head = cur = ListNode(0)
        while l1 or l2 or carry:
            s = (l1 and l1.val or 0) + (l2 and l2.val or 0) + carry
            carry = s / 10
            cur.next = ListNode(s % 10)
            cur = cur.next
            l1 = l1 and l1.next
            l2 = l2 and l2.next
        return head.next

File: algorithms/2. Add Two Numbers/solution.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        ListNode dummy(0), * cur = & dummy;
        while ( l1 || l2 || carry ) {
            carry += l1 ? l1->val : 0;
            carry += l2 ? l2->val : 0;
            l1 = l1 ? l1->next : l1;
            l2 = l2 ? l2->next : l2;
            cur->next = new ListNode(carry % 10);
            cur = cur->next;
            carry /= 10;
        }
        return dummy.next;
    }
};
